effekt/shared/src/main/scala/effekt/typer/PreTyper.scala

```diff
@@ -145,7 +159,7 @@ class BoxUnboxInference {
       VarDef(id, annot, region, rewrite(binding))
 
     case DefDef(id, annot, binding) =>
-      DefDef(id, annot, rewriteAsBlock(binding))
+      DefDef(id, annot, rewriteAsBlock(binding, Some(t)))
```

Comment:
Maybe you can run `rewriteAsBlock` first, then check whether an unbox was inserted, instead of passing the context parameter? The added parameter adds a bit confusion to the signature ("why is it necessary for the transformation to know about the `parentDef`?")

Not sure, whether analyzing and annotating at the point of `DefDef` is much clearer, though. What do you think?